src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs

```diff
@@ -1693,6 +1832,16 @@ private Task<(long Index, object Result)> UpdateTopology(UpdateTopologyCommand c
             return _engine.PutAsync(cmd);
         }
 
+        private Task<(long Index, object Result)> UpdateResponsibleNodeForTask(UpdateResponsibleNodeForTaskCommand command)
+        {
+            if (_engine.LeaderTag != _server.NodeTag)
+            {
+                throw new NotLeadingException("This node is no longer the leader, so we abort updating the responsible node for task");
+            }
```

Comment:
That is subject to race conditions, no?

Reply:
This is done in order to close the cluster observer thread as soon it isn't the leader.